storage/spanstore/ratelimit/rate_limit.go

```diff
@@ -16,36 +16,54 @@ package ratelimit
 
 import (
 	"errors"
+	"time"
 
+	"github.com/uber/jaeger-lib/metrics"
 	rlImpl "go.uber.org/ratelimit"
 
 	"github.com/jaegertracing/jaeger/model"
 	"github.com/jaegertracing/jaeger/storage/spanstore"
 )
 
-var errInvalidRate = errors.New("rate must be a non-zero positive integer")
+var errInvalidRate = errors.New("rate must be a positive integer")
+
+type rateLimitedWriter struct {
+	writer        spanstore.Writer
+	limiter       rlImpl.Limiter
+	sleepTime     metrics.Timer
+	delayedWrites metrics.Counter
+}
 
 // NewRateLimitedWriter decorates a Writer with a rate limiter in order to limit
 // the number of writes per second.
-func NewRateLimitedWriter(writer spanstore.Writer, rate int) (spanstore.Writer, error) {
+func NewRateLimitedWriter(
+	writer spanstore.Writer,
+	rate int,
+	metricsFactory metrics.Factory,
+) (spanstore.Writer, error) {
 	if rate <= 0 {
 		return nil, errInvalidRate
 	}
 
+	m := metricsFactory.Namespace("rate-limited-writer", nil)
 	return &rateLimitedWriter{
-		writer:  writer,
-		limiter: rlImpl.New(rate),
+		writer:        writer,
+		limiter:       rlImpl.New(rate),
+		sleepTime:     m.Timer("sleep-time", nil),
+		delayedWrites: m.Counter("delayed-writes", nil),
 	}, nil
 }
 
-type rateLimitedWriter struct {
-	writer  spanstore.Writer
-	limiter rlImpl.Limiter
-}
-
 // WriteSpan wraps the write span method of the inner writer, but invokes the
 // rate limiter before each write.
 func (r *rateLimitedWriter) WriteSpan(s *model.Span) error {
-	r.limiter.Take()
+	const threshold = 100 * time.Microsecond
```

Comment:
I happen to agree wholeheartedly. Team suggested I add a metric with number of delayed spans. The current rate limiter API makes this essentially impossible. IMO the real issue is that the rate limiter API has limited use in its current form. I believe it would have made more sense to return the duration slept as opposed to the timestamp of the wake time.